File: DSA/VivaswanSingh_590023022/Day_95/q2.c
#include <stdio.h>

int festivalSchedule(int T, int duration[], int happiness[], int n)
{
    int dp[T + 1];

    for (int t = 0; t <= T; t++)
        dp[t] = 0;

    for (int i = 0; i < n; i++)
    {
        for (int t = T; t >= duration[i]; t--)
        {
            int value = dp[t - duration[i]] + happiness[i];

            if (value > dp[t])
                dp[t] = value;
        }
    }

    return dp[T];
}

int main()
{
    int T = 7;

    int duration[] = {2, 3, 4, 5};
    int happiness[] = {4, 5, 7, 8};

    int n = sizeof(duration) / sizeof(duration[0]);

    int result = festivalSchedule(T, duration, happiness, n);

    printf("Maximum happiness: %d\n", result);

    return 0;
}

File: DSA/vaishnavi_joshi_590016947/Day95/Q2.java
class Solution {
    public int maxHappiness(int n, int T, int[] duration, int[] happiness) {

        // dp[t] = maximum happiness possible with t time
        int[] dp = new int[T + 1];

        for (int i = 0; i < n; i++) {

            // Traverse backwards for 0/1 Knapsack
            for (int t = T; t >= duration[i]; t--) {

                dp[t] = Math.max(
                    dp[t],
                    dp[t - duration[i]] + happiness[i]
                );
            }
        }

        return dp[T];
    }
}